### PR TITLE
chore(warehouse): validate object storage credentials

### PR DESCRIPTION
**Pull request by deepakrai9185720**
# Description

We store events that are flowing through our systems so that in case of any issues, we can replay them back and get at correct state. We store these events in bucket owned by rudderstack.

Going forward, we will be asking customers if they want to store the events on there own systems and if yes, allow them to add the credentials of the storage. The task is to expose a validation endpoint in warehouse to capture the creds and validate them on account of ability to upload and download.

The outcome of this validation will be presented to the user.

## Notion Ticket

https://www.notion.so/rudderstacks/Object-Storage-Validation-Design-0a3300c4754d4271bed53d9ad15d683d

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/api.go [hunk 1]
```diff
@@ -92,8 +96,14 @@ type UploadAPIT struct {
 	isMultiWorkspace  bool
 }
 
+type ValidateObjectRequestT struct {
```
Comment (deepakrai9185720) [inline warehouse/api.go:99]: Can be removed

File: warehouse/warehousegrpc.go [hunk 2]
```diff
@@ -88,3 +92,63 @@ func (w *warehousegrpc) RetryWHUploads(ctx context.Context, req *proto.RetryWHUp
...
+	}
+
+	r := ObjectStorageValidationRequest{}
+	if err := json.Unmarshal(byt, &r); err != nil {
+		return nil, status.Errorf(
+			codes.Code(code.Code_INTERNAL),
+			"unable to extract data into validation request with error: %s", err)
+	}
+
+	switch r.Type {
```
Comment (deepakrai9185720) [inline warehouse/warehousegrpc.go:117]: move this  into a new function

File: warehouse/api.go [hunk 3]
```diff
@@ -92,8 +96,14 @@ type UploadAPIT struct {
...
+	body string
+}
+
 var UploadAPI UploadAPIT
 
+var DownloadFileNamePattern = "downloadfile.*.tmp"
```
Comment (achettyiitr) [inline warehouse/api.go:105]: Can we make this constant?

File: warehouse/api.go (hunk 1 above)
Comment (achettyiitr) [inline warehouse/api.go:99]: If this is not used, Can we remove this?

File: warehouse/api.go [hunk 4]
```diff
@@ -611,3 +621,111 @@ func (uploadsReq *UploadsReqT) getWhUploads(selectFields string) (uploadsRes *pr
...
+		return false
+	}
+
+	if valStr, ok := value.(string); !ok {
+		return false
+	} else {
+		return len(valStr) != 0
+	}
+
+	return false
```
Comment (achettyiitr) [inline warehouse/api.go:640]: Unreachable statement.

File: warehouse/api.go [hunk 5]
```diff
@@ -611,3 +621,111 @@ func (uploadsReq *UploadsReqT) getWhUploads(selectFields string) (uploadsRes *pr
...
+	f, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("unable to open path to temporary file: %w", err)
+	}
+
+	uploadOutput, err := fileManager.Upload(ctx, f)
+	if err != nil {
+		return InvalidDestinationCredErr{Base: err, Operation: "upload"}
+	}
+	f.Close()
```
Comment (achettyiitr) [inline warehouse/api.go:673]: nit: linting issue.
`_ = f.Close()`

File: warehouse/api.go [hunk 6]
```diff
@@ -611,3 +621,111 @@ func (uploadsReq *UploadsReqT) getWhUploads(selectFields string) (uploadsRes *pr
...
+		return fmt.Errorf("error while Creating file to download data")
+	}
+
+	defer os.Remove(f.Name())
+
+	err = fileManager.Download(ctx, f, key)
+	if err != nil {
+		return InvalidDestinationCredErr{Base: err, Operation: "download"}
+	}
+	f.Close()
```
Comment (achettyiitr) [inline warehouse/api.go:693]: nit: linting issue.
`_ = f.Close()`

File: warehouse/api.go [hunk 7]
```diff
@@ -611,3 +621,111 @@ func (uploadsReq *UploadsReqT) getWhUploads(selectFields string) (uploadsRes *pr
 	}
 	return
 }
+
+// checkMapForValidKey checks the presence of key in map
+// and if yes verifies that the key is string and non-empty.
+func checkMapForValidKey(configMap map[string]interface{}, key string) bool {
```
Comment (achettyiitr) [inline warehouse/api.go:627]: Let's have unit Test for this alter.

File: warehouse/api.go [hunk 8]
```diff
@@ -611,3 +621,111 @@ func (uploadsReq *UploadsReqT) getWhUploads(selectFields string) (uploadsRes *pr
...
+		ifNotExistThenSet("accessKey", envConfig["accessKey"], settings.Config)
+		ifNotExistThenSet("enableSSE", envConfig["enableSSE"], settings.Config)
+		ifNotExistThenSet("iamRoleARN", envConfig["iamRoleArn"], settings.Config)
+		ifNotExistThenSet("externalID", envConfig["externalId"], settings.Config)
+		ifNotExistThenSet("regionHint", envConfig["regionHint"], settings.Config)
+	}
+}
+
+//
+func ifNotExistThenSet(keyToReplace string, replaceWith interface{}, configMap map[string]interface{}) {
```
Comment (achettyiitr) [inline warehouse/api.go:726]: Let's have unit Test for this alter.

File: warehouse/errors.go [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package warehouse
+
+import "fmt"
+
+type InvalidDestinationCredErr struct {
+	Base      error
+	Operation string
+}
+
+func (err InvalidDestinationCredErr) Error() string {
```
Comment (achettyiitr) [inline warehouse/errors.go:10]: This is really an excellent way to start exposing customer-specific errors.

File: warehouse/api.go [hunk 10]
```diff
@@ -611,3 +621,111 @@ func (uploadsReq *UploadsReqT) getWhUploads(selectFields string) (uploadsRes *pr
...
+		Config:   inputConfig,
+	}
+
+	overrideWithEnv(settings)
+	return settings
+}
+
+// overrideWithEnv overrides the config keys in the filemanager settings
+// with fallback values pulled from env. Only supported for S3 for now.
+func overrideWithEnv(settings *filemanager.SettingsT) {
```
Comment (achettyiitr) [inline warehouse/api.go:710]: Let's have a unit test for this afterwards.

File: warehouse/warehousegrpc.go [hunk 11]
```diff
@@ -88,3 +91,70 @@ func (w *warehousegrpc) RetryWHUploads(ctx context.Context, req *proto.RetryWHUp
...
+	Type   string                 `json:"type"`
+	Config map[string]interface{} `json:"config"`
+}
+
+func validateObjectStorageRequestBody(request *proto.ValidateObjectStorageRequest) (*ObjectStorageValidationRequest, error) {
+
+	byt, err := json.Marshal(request)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Code(code.Code_INTERNAL),
```
Comment (achettyiitr) [inline warehouse/warehousegrpc.go:105]: Nice to start using grpc error codes.

File: warehouse/warehousegrpc.go [hunk 12]
```diff
@@ -88,3 +91,70 @@ func (w *warehousegrpc) RetryWHUploads(ctx context.Context, req *proto.RetryWHUp
 	}
 	return
 }
+
+type ObjectStorageValidationRequest struct {
+	Type   string                 `json:"type"`
+	Config map[string]interface{} `json:"config"`
+}
+
+func validateObjectStorageRequestBody(request *proto.ValidateObjectStorageRequest) (*ObjectStorageValidationRequest, error) {
```
Comment (achettyiitr) [inline warehouse/warehousegrpc.go:100]: Let's have a test case for this afterwards.

File: warehouse/api.go (hunk 1 above)
Comment (deepakrai9185720) [reply warehouse/api.go]: Done

**Review by achettyiitr — APPROVED** [review]

**Comment by achettyiitr** [comment]
Depending upon the rudder server features settings.
https://github.com/rudderlabs/rudder-server/pull/2377

**Review by Jayachand — APPROVED** [review]

**Pull request merged**